src/testing/jest/jest-stencil-connector.ts

```diff
@@ -0,0 +1,114 @@
+/*!
+ * This file acts as the connector/bridge between Stencil and Jest.
+ *
+ * It defines/caches a `JestAdapter` instance to dispatch Jest-related configuration calls to the correct section of the
+ * Stencil codebase.
+ *
+ * It contains the APIs that are designed to be used by the Jest pre-configurations supplied by Stencil.
+ */
+
+import semverMajor from 'semver/functions/major';
+
+import { Jest27StencilAdapter } from './jest-27-and-under/jest-facade';
+import { JestAdapter } from './jest-adapter';
+import { getJestMajorVersion } from './jest-apis';
+
+/**
+ * Store a reference to the Jest version-specific adapter used to get pieces of testing infrastructure
+ */
+let JEST_ADAPTER: JestAdapter | null = null;
+
+/**
+ * Retrieve the numeric representation of the major version of Jest being used.
+ *
+ * If a user has Jest v27.1.0 installed, `27` will be returned.
+ *
+ * @returns the major version of Jest detected
+ */
+export const getVersion = (): number => {
+  return semverMajor(getJestMajorVersion());
+};
+
+/**
+ * Retrieve the cached local variable containing a Jest adapter, based on the version of Jest detected.
+ * If no Jest adapter is cached, set it.
+ *
+ * @returns the cached Jest adapter.
+ */
+const getJestAdapter = (): JestAdapter => {
+  if (!JEST_ADAPTER) {
+    const version = getVersion();
+    if (version <= 27) {
+      JEST_ADAPTER = new Jest27StencilAdapter();
```

Comment:
I thought maybe this would be able to resolve the "adapter" based on consistent file structure and naming conventions (and a default export from the file). That too magical?

Reply:
The original prototype actually used something very similar to what you're describing:

```ts
class Jest27StencilAdapter extends JestFacade {
  static override async getRunner() {
    return await import('./experimental/jest-27-and-under/jest-runner');
  }
  static override async getScreenshot() {
    return await import('./experimental/jest-27-and-under/jest-screenshot');
  }
}
```

I found there were a couple shortcomings with this:
1. Type safety wasn't as "nice" as I wanted it to be (the return type of `getRunner` is a `Promise<>` that returns the shape of _whatever_ the export of `jest-runner` is). I found this to be kind of annoying to work with/harder to understand
2. It introduced `async/await` to the the whole call chain, which wasn't strictly necessary
3. Refactoring (say for instance, I were to hoist 'jest-27-and-under' out of that 'experimental' directory 😉) wasn't very pleasurable either.

By using the Facade pattern in this way, we can be super explicit about what adapters/implementations map to which versions of Jest, avoid some async/await overhead, and have some better (IMO) refactoring ability.